torch/_inductor/fx_passes/post_grad.py

```diff
@@ -802,3 +805,162 @@ def repl(inp, x1, x2):
 
     with V.fake_mode:
         match.replace_by_example(repl, [inp, mat1, mat2])
+
+
+def allows_mixed_devices(op):
+    return op in (
+        aten.index.Tensor,
+        aten.index_put.default,
+        aten.index_put_.default,
+        aten.copy.default,
+        aten.copy_.default,
+        aten.slice_scatter.default,
+    )
+
+
+def cannot_be_moved_to_cuda(node):
+    if node.target == "output":
+        return True
+
+    if not isinstance(
+        node.target, torch._ops.OpOverload
+    ) and node.target.namespace not in ("prims", "aten"):
+        return True
+
+    # can take mixed devices - bail for now
+    incompatible_ops = (aten._embedding_bag.default,)
+
+    return node.target in incompatible_ops
+
+
+def get_node_device(node: fx.Node) -> Optional[torch.device]:
+    ten = node.meta.get("val")
+    return None if not isinstance(ten, torch.Tensor) else ten.device
+
+
+def get_cpu_indeg_count(graph) -> Dict[fx.Node, int]:
+    cpu_indeg: Dict[fx.Node, int] = Counter()
+
+    for node in graph.nodes:
+        cpu_count = 0
+
+        def add_cpu_inp(node):
+            nonlocal cpu_count
+            device = get_node_device(node)
+            cpu_count += device is not None and device.type == "cpu"
+
+        pytree.tree_map_only(torch.fx.Node, add_cpu_inp, (node.args, node.kwargs))
+
+        if cpu_count:
+            cpu_indeg[node] = cpu_count
+
+    return cpu_indeg
+
+
+def move_constructors_to_cuda(graph):
+    """
+    Moves intermediary tensors which are constructed on the cpu to cuda when safe
+    """
+    if not torch.backends.cuda.is_built():
+        return
+
+    cuda_devices = set()
+    constructors = []
+
+    for node in graph.nodes:
+        device = get_node_device(node)
+        if device and device.type == "cuda":
+            cuda_devices.add(device)
+
+        if not isinstance(
+            node.target, torch._ops.OpOverload
+        ) or node.target.namespace not in ("prims", "aten"):
+            continue
+
+        if not torch._subclasses.fake_tensor._is_tensor_constructor(node.target):
+            continue
+
+        if not node.kwargs.get("device") == torch.device("cpu"):
+            continue
+
+        constructors.append(node)
+
+    # not handling multiple cuda devices initially
+    if not constructors or len(cuda_devices) != 1:
+        return
+
+    movable_constructors = find_movable_constructors(graph, constructors)
+
+    for node in movable_constructors:
+        kwargs = node.kwargs.copy()
+        kwargs["device"] = next(iter(cuda_devices))
+        node.kwargs = kwargs
+
+
+def find_movable_constructors(graph, constructors: List[fx.Node]) -> Set[fx.Node]:
+    """
+    Starting from the cpu constructors, iterate through the graph and test that all of their
+    downstream uses can safely be moved to cpu.
```

Comment:
Maybe this is just a matter of taste, but to me it seems like it would make more sense to do this as a backwards pass: e.g., we start with the cpu-cuda transition barrier, and then we try to successively absorb cpu operations into the cuda region, so long as doing so wouldn't increase DtoH memory traffic.

Reply:
This is a good suggestion - for now i'm going to see how dashboard looks for current impl. Some of the other backends (xla) may want to be more aggressive and not include that.